es72x/src/main/scala/tech/beshu/ror/es/dlsfls/DocumentFieldReader.scala

```diff
@@ -0,0 +1,231 @@
+          case Some(name) => currentBuilder.startArray(name)
+          case None => currentBuilder.startArray()
+        }
+        json match {
+          case Obj(map) =>
+            map
+              .foldLeft(startObject(builder)) {
+                case (currentBuilder, (fieldName, fieldValue)) =>
+                  val newlyCollectedField = currentField(collectedField, fieldName)
+                  if(policy.canKeep(newlyCollectedField)) {
+                    buildFilteredResponse(currentBuilder, Some(fieldName), fieldValue, newlyCollectedField)
+                  } else {
+                    currentBuilder
+                  }
+              }
+              .endObject()
+          case Arr(values) =>
+            values
+              .foldLeft(startArray(builder)) {
+                case (currentBuilder, arrayObject) =>
+                  buildFilteredResponse(currentBuilder, None, arrayObject, collectedField)
+              }
+              .endArray()
+          case Str(value) =>
+            key match {
+              case Some(aKey) => builder.field(aKey, value)
+              case None => builder.value(value)
+            }
+          case Num(value) =>
+            key match {
+              case Some(aKey) => builder.field(aKey, value)
+              case None => builder.value(value)
+            }
+          case Bool(value) =>
+            key match {
+              case Some(aKey) => builder.field(aKey, value)
+              case None => builder.value(value)
+            }
+          case Null =>
+            key match {
+              case Some(aKey) => builder.nullField(aKey)
+              case None => builder.nullValue()
+            }
+        }
+      }
+
+      private def currentField(collectedField: String, currentFieldPart: String) = {
+        if(collectedField.isEmpty) currentFieldPart else s"$collectedField.$currentFieldPart"
+      }
+    })
+  }
+
+  override def getCoreCacheHelper: IndexReader.CacheHelper = this.in.getCoreCacheHelper
+
+  override def getReaderCacheHelper: IndexReader.CacheHelper = this.in.getCoreCacheHelper
+
+}
+
+object DocumentFieldReader {
+  def wrap(in: DirectoryReader, fields: NonEmptySet[DocumentField]): DocumentFieldDirectoryReader =
+    new DocumentFieldDirectoryReader(in, fields)
+}
+
+final class DocumentFieldDirectoryReader(in: DirectoryReader, fields: NonEmptySet[DocumentField])
+  extends FilterDirectoryReader(in, new DocumentFieldDirectorySubReader(fields)) {
+
+  override protected def doWrapDirectoryReader(in: DirectoryReader) =
+    new DocumentFieldDirectoryReader(in, fields)
+
+  override def getReaderCacheHelper: IndexReader.CacheHelper =
+    in.getReaderCacheHelper
+}
+
+object DocumentFieldDirectoryReader {
+  private class DocumentFieldDirectorySubReader(fields: NonEmptySet[DocumentField])
+    extends FilterDirectoryReader.SubReaderWrapper {
+
+    override def wrap(reader: LeafReader): LeafReader = {
+      Try(new DocumentFieldReader(reader, fields))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Usage of get on optional type.](https://app.codacy.com/manual/scarduzio/elasticsearch-readonlyrest-plugin/pullRequest?prid=4803411)